### PR TITLE
feat: add sqlite storage backend

### PR DESCRIPTION
**Pull request by hanxiao**
- [x] pass the test `test_advance_indexing.py`
- [x] sugar API on `DocumentArray` (what @alaeddine-13 proposed)
- [ ] have some basic sense of the speed. how much slowdown one should expect by using sqlite storage.

### CODE REVIEW

File: docarray/array/base.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    def __new__(cls, *args, storage: str = 'memory', **kwargs):
+        if cls is DocumentArray:
+            if storage == 'memory':
+                from docarray.array.memory import DocumentArrayInMemory
+                instance = super().__new__(DocumentArrayInMemory)
+            elif storage == 'sqlite':
+                from docarray.array.sqlite import DocumentArraySqlite
+                instance = super().__new__(DocumentArraySqlite)
+            else:
+                raise ValueError(f'storage=`{storage}` is not supported.')
```
Comment (numb3r3) [inline docarray/array/base.py:21]: `TypeError` would be better to indicate the error type.

File: docarray/array/base.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+from abc import ABC
+from docarray.array.mixins import AllMixins
+
+
+class BaseDocumentArray(ABC):
+    def __init__(self, *args, storage: str = 'memory', **kwargs):
+        super().__init__()
+        self._init_storage(*args, **kwargs)
```
Comment (numb3r3) [inline docarray/array/base.py:8]: Is  `_init_storage()` still necessary?

File: docarray/array/chunk.py [hunk 3]
```diff
@@ -7,12 +7,13 @@
...
+from .memory import DocumentArrayInMemory
 
 if TYPE_CHECKING:
     from ..document import Document
 
 
-class ChunkArray(DocumentArray):
+class ChunkArray(DocumentArrayInMemory):
     """
     :class:`ChunkArray` inherits from :class:`DocumentArray`.
```
Comment (numb3r3) [inline docarray/array/chunk.py:18]: please don't forget to update the doc str accordingly

File: docarray/array/base.py (hunk 1 above)
Comment (hanxiao) [reply docarray/array/base.py:21]: in this case it is `ValueError`

File: docarray/array/base.py (hunk 2 above)
Comment (hanxiao) [reply docarray/array/base.py:8]: yes

Comment (hanxiao) [reply docarray/array/base.py:8]: oh im actually not so sure now, will wait for @alaeddine-13

File: docarray/array/storage/sqlite/helper.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+def _do_create_table(
+    table_name: str,
+    cur: 'sqlite3.Cursor',
+) -> None:
+    cur.execute(
+        f'''
+            CREATE TABLE {table_name} (
+            doc_id TEXT NOT NULL UNIQUE, 
+            serialized_value Document NOT NULL, 
+            item_order INTEGER PRIMARY KEY)
```
Comment (hanxiao) [inline docarray/array/storage/sqlite/helper.py:47]: 1. in weaviate, `add(..., vector=PrefilledValue)`

```suggestion
            doc_id TEXT NOT NULL UNIQUE, 
            serialized_value Document NOT NULL, 
            item_order INTEGER PRIMARY KEY
            (+vector)
            )
```

Comment (hanxiao) [reply docarray/array/storage/sqlite/helper.py:47]: 2. override `.match()`

File: docarray/array/storage/sqlite/getsetdel.py [hunk 5]
```diff
@@ -26,8 +26,8 @@ def _set_doc_by_offset(self, offset: int, value: 'Document'):
 
     def _set_doc_by_id(self, _id: str, value: 'Document'):
         self._sql(
-            f'UPDATE {self._table_name} SET serialized_value=? WHERE doc_id=?',
-            (value, _id),
```
Comment (numb3r3) [inline docarray/array/storage/sqlite/getsetdel.py]: What's the consequence after: 

```
da['a'] = Document(id='c', text='c')
```

In your commit, `da['a']` is still available, and not updated, am I right?

**Pull request merged**